project-clone/main.go

```diff
@@ -0,0 +1,278 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+
+	dw "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/go-git/go-git/v5"
+	gitConfig "github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"sigs.k8s.io/yaml"
+
+	"github.com/devfile/devworkspace-operator/pkg/library/constants"
+	"github.com/devfile/devworkspace-operator/pkg/provision/metadata"
+)
+
+var (
+	projectsRoot              string
+	flattenedDevWorkspacePath string
+)
+
+func init() {
+	projectsRoot = os.Getenv(constants.ProjectsRootEnvVar)
+	if projectsRoot == "" {
+		log.Printf("Required environment variable %s is unset", constants.ProjectsRootEnvVar)
+	}
+
+	flattenedDevWorkspacePath = os.Getenv(metadata.FlattenedDevfileMountPathEnvVar)
+	if flattenedDevWorkspacePath == "" {
+		log.Printf("Required environment variable %s is unset", metadata.FlattenedDevfileMountPathEnvVar)
+	}
+}
+
+func main() {
+	workspace, err := readFlattenedDevWorkspace()
+	if err != nil {
+		log.Printf("Failed to read current DevWorkspace: %s", err)
+		os.Exit(1)
+	}
+	for _, project := range workspace.Projects {
+		needClone, needRemotes, err := checkProjectState(&project)
+		if err != nil {
+			log.Printf("Failed to process project %s: %s", project.Name, err)
+			os.Exit(1)
+		}
+		if needClone {
+			repo, err := cloneProject(&project)
+			if err != nil {
+				log.Printf("failed to clone project %s: %s", project.Name, err)
+				os.Exit(1)
+			}
+			if err := setupRemotes(repo, &project); err != nil {
+				log.Printf("Failed to set up remotes for project %s: %s", project.Name, err)
+				os.Exit(1)
+			}
+			if err := handleCheckoutFromReference(repo, &project); err != nil {
+				log.Printf("Failed to checkout revision for project %s: %s", project.Name, err)
+			}
+		} else if needRemotes {
+			repo, err := openRepo(&project)
+			if err != nil {
+				log.Printf("Failed to open existing project %s: %s", project.Name, err)
+			}
+			if err := setupRemotes(repo, &project); err != nil {
+				log.Printf("Failed to set up remotes for project %s: %s", project.Name, err)
+				os.Exit(1)
+			}
+		} else {
+			log.Printf("Nothing to do for project %s", project.Name)
+		}
+	}
+}
+
+func checkProjectState(project *dw.Project) (needClone, needRemotes bool, err error) {
```

Comment:
main.go is pretty big. I wonder if it makes sense to move lower-level logic to a separate file.

Reply:
Your wish is my command :D

Reply:
Only when you think it makes sense ;)